Neos.Flow/Classes/Session/Data/SessionMetaData.php

```diff
@@ -0,0 +1,140 @@
+     * @param string[] $tags
+     */
+    public function __construct(
+        public /** readonly */ string $sessionIdentifier,
+        public /** readonly */ string $storageIdentifier,
+        public /** readonly */ int $lastActivityTimestamp,
+        public /** readonly */ array $tags
+    ) {
+    }
+
+    public static function createWithTimestamp(int $timestamp): self
+    {
+        return new self(
+            Algorithms::generateRandomString(32),
+            Algorithms::generateUUID(),
+            $timestamp,
+            []
+        );
+    }
+
+    /**
+     * Create session metadata from classic cache format for backwards compatibility
+     * @param string $sessionIdentifier
+     * @param array{'storageIdentifier': string, 'lastActivityTimestamp': int, 'tags': string[]} $data
+     * @deprecated this will be removed with flow 10
+     */
+    public static function createFromSessionIdentifierAndOldArrayCacheFormat(string $sessionIdentifier, array $data): self
+    {
+        return new self(
+            $sessionIdentifier,
+            $data['storageIdentifier'],
+            $data['lastActivityTimestamp'],
+            $data['tags']
+        );
+    }
+
+    public function withLastActivityTimestamp(int $lastActivityTimestamp): self
+    {
+        return new self(
+            $this->sessionIdentifier,
+            $this->storageIdentifier,
+            $lastActivityTimestamp,
+            $this->tags
+        );
+    }
+
+    public function withNewSessionIdentifier(): self
+    {
+        return new self(
+            Algorithms::generateRandomString(32),
+            $this->storageIdentifier,
+            $this->lastActivityTimestamp,
+            $this->tags
+        );
+    }
+
+    public function withAddedTag(string $tag): self
+    {
+        $tags = $this->tags;
+        if (!in_array($tag, $this->tags)) {
+            $tags[] = $tag;
+        }
+        return new self(
+            $this->sessionIdentifier,
+            $this->storageIdentifier,
+            $this->lastActivityTimestamp,
+            $tags
+        );
+    }
+
+    public function withRemovedTag(string $tag): self
+    {
+        $tags = $this->tags;
+        $index = array_search($tag, $tags);
+        if ($index !== false) {
+            unset($tags[$index]);
+        }
+        return new self(
+            $this->sessionIdentifier,
+            $this->storageIdentifier,
+            $this->lastActivityTimestamp,
+            $tags
+        );
+    }
+
+    /**
+     * Determine whether the metadata is equal in all aspects other than lastActivityTimestamp
+     */
+    public function isSame(SessionMetaData $other): bool
+    {
+        if ($this->sessionIdentifier != $other->sessionIdentifier) {
+            return false;
+        }
+
+        if ($this->storageIdentifier != $other->storageIdentifier) {
+            return false;
+        }
+
+        if (array_diff($this->tags, $other->tags) || array_diff($other->tags, $this->tags)) {
```

Comment:
this seems not super optimised ...

Reply:
adjusted via. https://github.com/neos/flow-development-collection/pull/3254/commits/1a16e7e075d9ad3e3320a13da73ddf25355e18f4